dali/kernels/signal/fft/fft_cpu_impl_utils.h

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_FFT_FFT_CPU_IMPL_UTILS_H_
+#define DALI_KERNELS_SIGNAL_FFT_FFT_CPU_IMPL_UTILS_H_
+
+#include <utility>
+#include <vector>
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace fft {
+namespace impl {
+
+inline int64_t next_pow2(int64_t n) {
+  int64_t pow2 = 2;
+  while (n > pow2) {
+    pow2 *= 2;
+  }
+  return pow2;
+}
+
+inline bool is_pow2(int64_t n) {
+  return (n & (n-1)) == 0;
+}
+
+inline bool can_use_real_impl(int64_t n) {
+  // Real impl can be selected when doing forward transform and n is a power of 2
+  return is_pow2(n);
+}
+
+inline int64_t size_in_buf(int64_t n) {
+  // Real impl input needs:    N real numbers    -> N floats
+  // Complex impl input needs: N complex numbers -> 2*N floats
+  return can_use_real_impl(n) ? n : 2*n;
+}
+
+inline int64_t size_out_buf(int64_t n) {
+  // Real impl output needs:    (N/2)+1 complex numbers -> N+2 floats
+  // Complex impl output needs: N complex numbers       -> 2*N floats
+  return can_use_real_impl(n) ? n+2 : 2*n;
+}
+
+struct ComplexSpectrumCalculator {
+  template <typename OutputType = std::complex<float>, typename InputType = std::complex<float>>
+  void Calculate(OutputType *out, const InputType *in,
+                 int64_t nfft, int64_t out_stride = 1, int64_t in_stride = 1) {
+    for (int i = 0; i <= nfft / 2; i++) {
+      out[i*out_stride] = in[i*in_stride];
+    }
+
+    for (int i = nfft / 2 + 1; i < nfft; i++) {
+      // start mirroring nfft/2+1+i -> nfft/2-1-i
+      auto tmp = in[(nfft - i)*in_stride];
+      out[i*out_stride] = {tmp.real(), -tmp.imag()};
+    }
+  }
+};
+
+struct MagnitudeSpectrumCalculator {
+  template <typename OutputType = float, typename InputType = std::complex<float>>
+  void Calculate(FftSpectrumType spectrum_type, OutputType *out, const InputType *in,
+                 int64_t nfft, int64_t out_stride = 1, int64_t in_stride = 1) {
```

Comment:
This needlessly binds this implementation to FFT - it should just tak input size and don't care about n/2 thing.

Reply:
will change